backend/supabase/migrations/20240414161927_users.sql

```diff
@@ -0,0 +1,276 @@
+        USING (TRUE);
+
+-------------------------------------------------------
+-- Section - Views
+-------------------------------------------------------
+
+-- TODO
+
+-------------------------------------------------------
+-- Section - RPC Functions
+-------------------------------------------------------
+
+CREATE OR REPLACE FUNCTION public.get_user_id(account_name citext)
+    RETURNS uuid
+    SECURITY DEFINER
+    LANGUAGE plpgsql
+    STABLE
+    AS $$
+BEGIN
+    RETURN(
+        SELECT
+            u.id
+        FROM
+            users u
+        WHERE
+            u.account_name = get_user_id.account_name);
+END;
+$$;
+
+GRANT EXECUTE ON FUNCTION public.get_user_id(citext) TO authenticated, service_role;
+
+CREATE OR REPLACE FUNCTION public.get_user_by_id(user_id uuid)
+    RETURNS public.users
+    SET search_path = public
+    SECURITY DEFINER
+    LANGUAGE plpgsql
+    STABLE
+    AS $$
+BEGIN
+    RETURN u
+FROM
+    users u
+WHERE
+    u.id = get_user_by_id.user_id;
+END;
+$$;
+
+GRANT EXECUTE ON FUNCTION public.get_user_by_id(uuid) TO authenticated, service_role;
+
+CREATE OR REPLACE FUNCTION public.get_user_by_name(account_name citext)
+    RETURNS public.users
+    SET search_path = public
+    SECURITY DEFINER
+    LANGUAGE plpgsql
+    STABLE
+    AS $$
+BEGIN
+    RETURN u
+FROM
+    users u
+WHERE
+    u.account_name = get_user_by_name.account_name;
+END;
+$$;
+
+GRANT EXECUTE ON FUNCTION public.get_user_by_name(citext) TO authenticated, service_role;
+
+CREATE OR REPLACE FUNCTION public.get_me()
+    RETURNS public.users
+    SET search_path = public
+    SECURITY DEFINER
+    LANGUAGE plpgsql
+    STABLE
+    AS $$
+BEGIN
+    RETURN u
+FROM
+    users u
+WHERE
+    u.id = auth.uid();
+END;
+$$;
+
+GRANT EXECUTE ON FUNCTION public.get_me() TO authenticated;
+
+CREATE OR REPLACE FUNCTION public.update_user(user_id uuid, display_name text DEFAULT NULL, bio text DEFAULT NULL, public_metadata jsonb DEFAULT NULL, replace_metadata boolean DEFAULT FALSE)
+    RETURNS public.users
+    SET search_path = public
+    SECURITY DEFINER
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    UPDATE
+        public.users u
+    SET
+        display_name = coalesce(update_user.display_name, u.display_name),
+        bio = coalesce(update_user.bio, u.bio),
+        public_metadata = CASE WHEN update_user.public_metadata IS NULL THEN
+            u.public_metadata
+        WHEN u.public_metadata IS NULL THEN
+            update_user.public_metadata
+        WHEN update_user.replace_metadata THEN
+            update_user.public_metadata
+        ELSE
+            u.public_metadata || update_user.public_metadata
+        END
+    WHERE
+        u.id = update_user.user_id;
+    RETURN get_user_by_id(update_user.user_id);
+END;
+$$;
+
+GRANT EXECUTE ON FUNCTION public.update_user(uuid, text, text, jsonb, boolean) TO authenticated, service_role;
+
+CREATE OR REPLACE FUNCTION public.search_users(account_name citext DEFAULT NULL)
```

Comment:
This allows for searching all users. And is public. This allows a full scan of all user. My understanding is that this will be restricted based on RLS @hitaspdotnet ?

Reply:
based on `users_select_policy` all authenticated users can select each other. just select. same as when you search a user in github to send invitation. 
I'll change return type to returns only account_name and ID or a view when views was ready.